infoview/goal.tsx

```diff
@@ -0,0 +1,66 @@
+import * as React from 'react';
+import { escapeHtml } from './util';
+import { ConfigContext } from './index';
+import { PlainGoal } from '../src/infoviewApi';
+
+interface GoalProps {
+    plainGoals: PlainGoal;
+}
+
+function emphasizeMessage(goal: string): string {
+    return goal
+        .replace(/([^`~@$%^&*()-=+\[{\]}⟨⟩⦃⦄⟦⟧⟮⟯‹›\\|;:",.\/\s]+)✝([¹²³⁴-⁹⁰]*)/g, '<span class="goal-inaccessible">$1$2</span>')
+        .replace(/^(⊢) /mg, '<strong class="goal-vdash">$1</strong> ')
+        .replace(/^(case) /mg, '<strong class="goal-case">$1</strong> ')
+        .replace(/^([^:\n< ][^:\n⊢{[(⦃]*) :/mg, '<strong class="goal-hyp">$1</strong> :')
+}
+
+export function Goal({plainGoals}: GoalProps): JSX.Element {
+    const config = React.useContext(ConfigContext);
+    const reFilters = config.infoViewTacticStateFilters || [];
+    const [filterIndex, setFilterIndex] = React.useState<number>(config.filterIndex ?? -1);
+    if (!plainGoals) return null;
+
+    let goals = plainGoals.goals;
+    if (!goals) {
+        goals = [];
+        const r = /```lean\n([^`]*)```/g
+        let match: RegExpExecArray
+        const unformatted = plainGoals.rendered;
+        do {
+            match = r.exec(unformatted)
+            if (match) {
+                goals.push(match[1])
+            }
+        } while (match)
+    }
+
+    if (!(reFilters.length === 0 || filterIndex === -1)) {
+        goals = goals.map((g) =>
+            // this regex splits the goal state into (possibly multi-line) hypothesis and goal blocks
+            // by keeping indented lines with the most recent non-indented line
+            g.match(/(^(?!  ).*\n?(  .*\n?)*)/mg).map((line) => line.trim())
```

Comment:
Should we make `goals` even more structured? For example, I thought about using the case name, if available, as the heading. We should also expose the pre-state while we're at it, so maybe `goals` isn't the best name in the first place.

Reply:
Yeah, the benefit would be negligible. By "pre-state" I meant the goals before execution of the tactic. Currently `goals` is always the post-state.